core/services/postgres/q.go

```diff
@@ -120,26 +120,46 @@ func (q Q) Transaction(lggr logger.Logger, fc func(q Queryer) error) error {
 	defer cancel()
 	return SqlxTransaction(ctx, q.Queryer, lggr, fc)
 }
-func (q Q) Query(query string, args ...interface{}) (*sql.Rows, error) {
+
+// CAUTION: A subtle problem lurks here, because the following code is buggy:
+//
+// ctx, cancel := context.WithCancel(context.Background())
+// rows, err := db.QueryContext(ctx, "SELECT foo")
+// cancel() // canceling here "poisons" the scan below
+// for rows.Next() {
+//   rows.Scan(...)
+// }
+//
+// We must cancel the context only after we have completely finished using the
+// returned rows or result from the query/exec
+//
+// For this reasons, the following functions return a context.CancelFunc and it
+// is up to the caller to ensure that cancel is called after it has finished
+//
+// Generally speaking, it makes more sense to use Get/Select in most cases anyway
+func (q Q) CtxQuery(query string, args ...interface{}) (*sql.Rows, context.CancelFunc, error) {
 	ctx, cancel := q.Context()
-	defer cancel()
-	return q.Queryer.QueryContext(ctx, query, args...)
+	rows, err := q.Queryer.QueryContext(ctx, query, args...)
+	return rows, cancel, err
 }
-func (q Q) Queryx(query string, args ...interface{}) (*sqlx.Rows, error) {
+func (q Q) CtxQueryx(query string, args ...interface{}) (*sqlx.Rows, context.CancelFunc, error) {
 	ctx, cancel := q.Context()
-	defer cancel()
-	return q.Queryer.QueryxContext(ctx, query, args...)
+	rows, err := q.Queryer.QueryxContext(ctx, query, args...)
+	return rows, cancel, err
 }
-func (q Q) QueryRowx(query string, args ...interface{}) *sqlx.Row {
+func (q Q) CtxQueryRowx(query string, args ...interface{}) (*sqlx.Row, context.CancelFunc) {
 	ctx, cancel := q.Context()
-	defer cancel()
-	return q.Queryer.QueryRowxContext(ctx, query, args...)
+	row := q.Queryer.QueryRowxContext(ctx, query, args...)
+	return row, cancel
 }
-func (q Q) Exec(query string, args ...interface{}) (sql.Result, error) {
+func (q Q) CtxExec(query string, args ...interface{}) (sql.Result, context.CancelFunc, error) {
```

Comment:
What do you think about adding a helper for the cases that don't use the result, so they only have to handle the error?
Also, `Ctx` in the name is a little confusing to me. What about this case being an `*Iter` or `*Result` variant of the simpler case?
```suggestion
func (q Q) Exec(query string, args ...interface{}) error {
  _, cancel, err := ExecIter(...)
  cancel()
  return err
}
func (q Q) ExecIter(query string, args ...interface{}) (sql.Result, context.CancelFunc, error) {
```

Reply:
Yup fair point, check latest changes